File: golang/priorityOrderExample/watch_priority_mempool.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/klauspost/compress/zstd"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/keepalive"
	"google.golang.org/grpc/metadata"

	pb "github.com/example/hyperliquid-grpc/proto"
)

const (
	defaultGrpcEndpoint = "your-endpoint.hype-mainnet.quiknode.pro:10000"
	defaultAuthToken    = "YOUR_QUICKNODE_TOKEN"
)

var zstdMagic = []byte{0x28, 0xB5, 0x2F, 0xFD}

func envOrDefault(name string, fallback string) string {
	value := os.Getenv(name)
	if value == "" {
		return fallback
	}
	return value
}

func createConnection(endpoint string) (*grpc.ClientConn, error) {
	return grpc.Dial(
		endpoint,
		grpc.WithTransportCredentials(credentials.NewClientTLSFromCert(nil, "")),
		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(100*1024*1024)),
		grpc.WithKeepaliveParams(keepalive.ClientParameters{
			Time:                30 * time.Second,
			Timeout:             10 * time.Second,
			PermitWithoutStream: true,
		}),
	)
}

func decompress(data []byte) (string, error) {
	if len(data) >= 4 &&
		data[0] == zstdMagic[0] &&
		data[1] == zstdMagic[1] &&
		data[2] == zstdMagic[2] &&
		data[3] == zstdMagic[3] {
		decoder, err := zstd.NewReader(nil)
		if err != nil {
			return "", err
		}
		defer decoder.Close()

		decompressed, err := decoder.DecodeAll(data, nil)
		if err != nil {
			return "", err
		}
		return string(decompressed), nil
	}

	return string(data), nil
}

func priorityFees(value interface{}) []string {
	fees := []string{}
	switch v := value.(type) {
	case map[string]interface{}:
		if _, ok := v["source"]; ok {
			if eventType, _ := v["type"].(string); eventType == "order" && v["p"] != nil {
				fees = append(fees, fmt.Sprint(v["p"]))
			}
		}
		if grouping, ok := v["grouping"].(map[string]interface{}); ok {
			if p, ok := grouping["p"]; ok {
				fees = append(fees, fmt.Sprint(p))
			}
		}
		for _, item := range v {
			fees = append(fees, priorityFees(item)...)
		}
	case []interface{}:
		for _, item := range v {
			fees = append(fees, priorityFees(item)...)
		}
	}
	return fees
}

func matchesTextFilters(text string, contains []string) bool {
	if len(contains) == 0 {
		return true
	}
	for _, needle := range contains {
		if strings.Contains(text, needle) {
			return true
		}
	}
	return false
}

func main() {
	startBlock := flag.Uint64("start-block", 0, "Start block for the stream")
	containsRaw := flag.String("contains", "", "Comma-separated text filters")
	includeConfirmed := flag.Bool("include-confirmed", false, "Also include confirmed ORDER_PRIORITY events from replica_cmds")
	rawMempoolFlag := flag.Bool("raw-mempool", false, "Subscribe to raw MEMPOOL_TXS and detect grouping.p locally")
	allMempool := flag.Bool("all-mempool", false, "With -raw-mempool, print all MEMPOOL_TXS messages, not only priority transactions")
	maxMessages := flag.Int("max-messages", 0, "Stop after printing this many matching messages")
	compact := flag.Bool("compact", false, "Print compact payload text instead of pretty JSON")
	flag.Parse()
	rawMempool := *rawMempoolFlag || *allMempool

	endpoint := envOrDefault("GRPC_ENDPOINT", defaultGrpcEndpoint)
	authToken := envOrDefault("AUTH_TOKEN", envOrDefault("QN_AUTH_TOKEN", defaultAuthToken))
	if endpoint == defaultGrpcEndpoint {
		log.Fatal("Set GRPC_ENDPOINT to your QuickNode Hyperliquid mainnet or testnet gRPC endpoint")
	}
	if authToken == defaultAuthToken {
		log.Fatal("Set AUTH_TOKEN to your QuickNode token")
	}

	contains := []string{}
	if *containsRaw != "" {
		for _, part := range strings.Split(*containsRaw, ",") {
			part = strings.TrimSpace(part)
			if part != "" {
				contains = append(contains, part)
			}
		}
	}

	conn, err := createConnection(endpoint)
	if err != nil {
		log.Fatalf("failed to connect: %v", err)
	}
	defer conn.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	ctx = metadata.AppendToOutgoingContext(ctx, "x-token", authToken)

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		cancel()
	}()

	client := pb.NewStreamingClient(conn)
	stream, err := client.StreamData(ctx)
	if err != nil {
		log.Fatalf("failed to create stream: %v", err)
	}

	streamType := pb.StreamType_ORDER_PRIORITY
	filters := map[string]*pb.FilterValues{
		"source": {Values: []string{"mempool_txs"}},
	}
	if rawMempool {
		streamType = pb.StreamType_MEMPOOL_TXS
		filters = map[string]*pb.FilterValues{}
	} else if *includeConfirmed {
		filters = map[string]*pb.FilterValues{}
	}

	if err := stream.Send(&pb.SubscribeRequest{
		Request: &pb.SubscribeRequest_Subscribe{
			Subscribe: &pb.StreamSubscribe{
				StreamType: streamType,
				StartBlock: *startBlock,
				Filters:    filters,
			},
		},
	}); err != nil {
		log.Fatalf("failed to subscribe: %v", err)
	}

	go func() {
		ticker := time.NewTicker(30 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				timestamp := time.Now().UnixMilli()
				fmt.Printf("PING timestamp=%d\n", timestamp)
				_ = stream.Send(&pb.SubscribeRequest{
					Request: &pb.SubscribeRequest_Ping{
						Ping: &pb.Ping{Timestamp: timestamp},
					},
				})
			}
		}
	}()

	if rawMempool {
		fmt.Println("Watching raw MEMPOOL_TXS")
	} else if *includeConfirmed {
		fmt.Println("Watching ORDER_PRIORITY events from mempool_txs and replica_cmds")
	} else {
		fmt.Println("Watching pre-consensus ORDER_PRIORITY mempool events")
	}
	fmt.Printf("Endpoint: %s\n", endpoint)
	if !rawMempool && !*includeConfirmed {
		fmt.Println("Server filter: source=mempool_txs (not finalized)")
	} else if rawMempool && !*allMempool {
		fmt.Println("Local filter: priority grouping only")
	}
	if len(contains) > 0 {
		fmt.Printf("Text filters: %v\n", contains)
	}

	printed := 0
	for {
		resp, err := stream.Recv()
		if err == io.EOF {
			return
		}
		if err != nil {
			log.Fatalf("receive error: %v", err)
		}
		if pong := resp.GetPong(); pong != nil {
			fmt.Printf("PONG timestamp=%d\n", pong.Timestamp)
			continue
		}

		dataUpdate, ok := resp.Update.(*pb.SubscribeUpdate_Data)
		if !ok {
			continue
		}

		text, err := decompress([]byte(dataUpdate.Data.Data))
		if err != nil {
			log.Printf("decompress error at block %d: %v", dataUpdate.Data.BlockNumber, err)
			continue
		}
		if !matchesTextFilters(text, contains) {
			continue
		}

		var parsed interface{}
		parsedOK := json.Unmarshal([]byte(text), &parsed) == nil
		fees := []string{}
		if parsedOK {
			fees = priorityFees(parsed)
		}
		if rawMempool && !*allMempool && len(fees) == 0 {
			continue
		}

		printed++
		fmt.Printf("\nBlock %d | Timestamp %d\n", dataUpdate.Data.BlockNumber, dataUpdate.Data.Timestamp)
		if len(fees) > 0 {
			fmt.Printf("Priority p: %s\n", strings.Join(fees, ", "))
		}
		if *compact {
			if len(text) > 1000 {
				fmt.Println(text[:1000])
			} else {
				fmt.Println(text)
			}
		} else if parsedOK {
			pretty, _ := json.MarshalIndent(parsed, "", "  ")
			fmt.Println(string(pretty))
		} else {
			fmt.Println(text)
		}

		if *maxMessages > 0 && printed >= *maxMessages {
			return
		}
	}
}

File: rust/src/priorityOrderExample/watch_priority_mempool.rs
use clap::Parser;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use tonic::transport::{Channel, ClientTlsConfig};
use tonic::{metadata::MetadataValue, Request};

pub mod hyperliquid {
    tonic::include_proto!("hyperliquid");
}

use hyperliquid::{
    streaming_client::StreamingClient, Ping, StreamSubscribe, StreamType,
    SubscribeRequest,
};

const DEFAULT_GRPC_ENDPOINT: &str = "https://your-endpoint.hype-testnet.quiknode.pro:10000";
const DEFAULT_AUTH_TOKEN: &str = "YOUR_QUICKNODE_TOKEN";
const ZSTD_MAGIC: [u8; 4] = [0x28, 0xB5, 0x2F, 0xFD];

fn grpc_endpoint() -> String {
    std::env::var("GRPC_ENDPOINT").unwrap_or_else(|_| DEFAULT_GRPC_ENDPOINT.to_string())
}

fn auth_token() -> String {
    std::env::var("AUTH_TOKEN")
        .or_else(|_| std::env::var("QN_AUTH_TOKEN"))
        .unwrap_or_else(|_| DEFAULT_AUTH_TOKEN.to_string())
}

async fn create_channel() -> Result<Channel, Box<dyn std::error::Error>> {
    let channel = Channel::from_shared(grpc_endpoint())?
        .tls_config(ClientTlsConfig::new())?
        .connect()
        .await?;
    Ok(channel)
}

fn decompress(data: &[u8]) -> Result<String, Box<dyn std::error::Error>> {
    if data.len() >= 4 && data[0..4] == ZSTD_MAGIC {
        let decompressed = zstd::decode_all(data)?;
        return Ok(String::from_utf8(decompressed)?);
    }

    Ok(String::from_utf8_lossy(data).to_string())
}

fn priority_fees(value: &serde_json::Value) -> Vec<String> {
    let mut fees = Vec::new();
    match value {
        serde_json::Value::Object(map) => {
            if let Some(grouping) = map.get("grouping") {
                if let Some(p) = grouping.get("p") {
                    fees.push(p.to_string());
                }
            }
            for item in map.values() {
                fees.extend(priority_fees(item));
            }
        }
        serde_json::Value::Array(items) => {
            for item in items {
                fees.extend(priority_fees(item));
            }
        }
        _ => {}
    }
    fees
}

fn matches_text_filters(text: &str, contains: &[String]) -> bool {
    contains.is_empty() || contains.iter().any(|needle| text.contains(needle))
}

#[derive(Parser)]
#[command(name = "priority-order-example")]
#[command(about = "Watch testnet priority MEMPOOL_TXS from a QuickNode Hyperliquid gRPC endpoint")]
struct Args {
    #[arg(long, default_value_t = 0)]
    start_block: u64,

    #[arg(long)]
    contains: Vec<String>,

    #[arg(long)]
    all_mempool: bool,

    #[arg(long)]
    max_messages: Option<usize>,

    #[arg(long)]
    compact: bool,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    if grpc_endpoint() == DEFAULT_GRPC_ENDPOINT {
        eprintln!("Set GRPC_ENDPOINT to your QuickNode Hyperliquid testnet gRPC endpoint");
        std::process::exit(2);
    }
    if auth_token() == DEFAULT_AUTH_TOKEN {
        eprintln!("Set AUTH_TOKEN to your QuickNode token");
        std::process::exit(2);
    }

    let channel = create_channel().await?;
    let mut client = StreamingClient::new(channel);
    let (tx, rx) = mpsc::channel(32);
    let stream = ReceiverStream::new(rx);

    tx.send(SubscribeRequest {
        request: Some(hyperliquid::subscribe_request::Request::Subscribe(
            StreamSubscribe {
                stream_type: StreamType::MempoolTxs as i32,
                start_block: args.start_block,
                filters: std::collections::HashMap::new(),
                filter_name: String::new(),
            },
        )),
    })
    .await?;

    let tx_ping = tx.clone();
    tokio::spawn(async move {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(30));
        loop {
            interval.tick().await;
            let _ = tx_ping
                .send(SubscribeRequest {
                    request: Some(hyperliquid::subscribe_request::Request::Ping(Ping {
                        timestamp: chrono::Utc::now().timestamp_millis(),
                    })),
                })
                .await;
        }
    });

    let mut request = Request::new(stream);
    request
        .metadata_mut()
        .insert("x-token", auth_token().parse::<MetadataValue<_>>()?);

    println!("Watching testnet MEMPOOL_TXS");
    println!("Endpoint: {}", grpc_endpoint());
    if !args.all_mempool {
        println!("Filter: priority grouping only");
    }
    if !args.contains.is_empty() {
        println!("Text filters: {:?}", args.contains);
    }

    let mut response_stream = client.stream_data(request).await?.into_inner();
    let mut printed = 0usize;

    while let Some(response) = response_stream.message().await? {
        let Some(update) = response.update else {
            continue;
        };
        let hyperliquid::subscribe_update::Update::Data(data) = update else {
            continue;
        };

        let text = decompress(data.data.as_bytes())?;
        if !matches_text_filters(&text, &args.contains) {
            continue;
        }

        let parsed = serde_json::from_str::<serde_json::Value>(&text).ok();
        let fees = parsed.as_ref().map(priority_fees).unwrap_or_default();
        if !args.all_mempool && fees.is_empty() {
            continue;
        }

        printed += 1;
        println!("\nBlock {} | Timestamp {}", data.block_number, data.timestamp);
        if !fees.is_empty() {
            println!("Priority fee grouping p: {}", fees.join(", "));
        }
        if args.compact {
            println!("{}", text.chars().take(1000).collect::<String>());
        } else if let Some(parsed) = parsed {
            println!("{}", serde_json::to_string_pretty(&parsed)?);
        } else {
            println!("{}", text);
        }

        if let Some(max) = args.max_messages {
            if printed >= max {
                return Ok(());
            }
        }
    }

    Ok(())
}
